x/configurl/dns.go

```diff
@@ -27,16 +28,52 @@ import (
 	"golang.org/x/net/dns/dnsmessage"
 )
 
-func wrapStreamDialerWithDO53(innerSD func() (transport.StreamDialer, error), innerPD func() (transport.PacketDialer, error), configURL *url.URL) (transport.StreamDialer, error) {
-	sd, err := innerSD()
-	if err != nil {
-		return nil, err
-	}
-	pd, err := innerPD()
-	if err != nil {
-		return nil, err
-	}
-	query := configURL.Opaque
+func registerDO53StreamDialer(r TypeRegistry[transport.StreamDialer], typeID string, newSD BuildFunc[transport.StreamDialer], newPD BuildFunc[transport.PacketDialer]) {
+	r.RegisterType(typeID, func(ctx context.Context, config *Config) (transport.StreamDialer, error) {
+		if config == nil {
+			return nil, fmt.Errorf("emtpy do53 config")
+		}
+		sd, err := newSD(ctx, config.BaseConfig)
+		if err != nil {
+			return nil, err
+		}
+		if closer, ok := sd.(io.Closer); ok {
+			defer closer.Close()
```

Comment:
Why are we closing the `sd` here? It will be passed to `dns.NewStreamDialer` below, will it be invalid after closing?

I think we may just need to close it when we return a non-`nil` error.

Reply:
Removed.

I was playing with the concept of cleaning up. We will need to add some mechanism for cleanup, or else we might leak resources. But I'll leave that to a future PR.